spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,89 +113,251 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
+			}
+		}
+		return false;
+	}
+
+	private boolean isDeclarativeMethodParameter(Object targetBean, MethodParameter methodParameter) {
+		if (targetBean != null) {
+			if (methodParameter.getParameterType().isAssignableFrom(targetBean.getClass())) {
+				return true;
+			}
+			for (StreamListenerParameterAdapter<?, Object> streamListenerParameterAdapter : streamListenerParameterAdapters) {
+				if (streamListenerParameterAdapter.supports(targetBean.getClass(), methodParameter)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	private String getInboundElementNameFromMethodAnnotation(StreamListener streamListener) {
+		return StringUtils.hasText(streamListener.value()) ? streamListener.value() : null;
+	}
+
+	private String getOutboundElementNameFromMethodAnnotation(Method method) {
+		SendTo sendTo = AnnotationUtils.findAnnotation(method, SendTo.class);
+		if (sendTo != null) {
+			Assert.isTrue(!ObjectUtils.isEmpty(sendTo.value()), "At least one output must be specified");
+			Assert.isTrue(sendTo.value().length == 1, "Multiple destinations cannot be specified");
+			Assert.hasText(sendTo.value()[0], "An empty destination cannot be specified");
+			return StringUtils.hasText(sendTo.value()[0]) ? sendTo.value()[0] : null;
+		}
+		Output output = AnnotationUtils.findAnnotation(method, Output.class);
+		if (output != null) {
+			Assert.isTrue(StringUtils.hasText(output.value()), "At least one output must be specified");
+			return StringUtils.hasText(output.value()) ? output.value() : null;
+		}
+		return null;
+	}
+
+	private void assertStreamListenerMethod(int inputAnnotationCount, int outputAnnotationCount,
+			String methodAnnotatedInboundName, String methodAnnotatedOutboundName, boolean isDeclarative) {
+		if (!isDeclarative) {
+			Assert.isTrue(inputAnnotationCount == 0 && outputAnnotationCount == 0,
+					"@Input or @Output annotation is not supported as method parameter in StreamListener method with " +
+							"message handler mapping");
+		}
+		if (StringUtils.hasText(methodAnnotatedInboundName) && StringUtils.hasText(methodAnnotatedOutboundName)) {
+			Assert.isTrue(inputAnnotationCount == 0 && outputAnnotationCount == 0, "@Input or @Output annotations are not permitted as " +
+					"method parameters when both inbound and outbound values are set as method annotated values");
+		}
+		if (StringUtils.hasText(methodAnnotatedInboundName)) {
+			Assert.isTrue(inputAnnotationCount == 0, "Cannot set both StreamListener value " + methodAnnotatedInboundName +
+					" and @Input annotation as method parameter");
+			Assert.isTrue(outputAnnotationCount == 0, "Cannot set StreamListener value '" + methodAnnotatedInboundName + "' when using" +
+					" @Output annotation as method parameter. Use @Input method parameter annotation to specify inbound value instead");
+		}
+		else {
+			Assert.isTrue(inputAnnotationCount >= 1, "No input destination is configured. Use either a @StreamListener argument or @Input");
```

Comment:
I believe technically it should be referred to as a "@StreamListener *attribute*"